.github/workflows/test-environment.yml

```diff
@@ -0,0 +1,102 @@
+name: sanity workflow
+
+on:
+  # Ability to execute on demand
+  workflow_dispatch:
+    inputs:
+      elk-stack-version:
+        required: true
+        description: "Elastic Cloud stack version"
+        default: "8.8.0-SNAPSHOT"
+      ess-region:
+        required: true
+        description: "Elastic Cloud deployment region"
+        default: "gcp-us-west2"
+
+env:
+  AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
+  AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+  AWS_REGION: "eu-west-3"
+  TF_VAR_ec_api_key: ${{ secrets.CI_EC_API_KEY }}
+  EC_DEPLOYMENT: "module.ec_deployment"
+  WORKING_DIR: deploy/cloud
+
+jobs:
+  Setup:
+    runs-on: ubuntu-20.04
+    timeout-minutes: 15
+    outputs:
+      stack-version: ${{ steps.set-elk-vars.outputs.stack-version }}
+      ess-region: ${{ steps.set-elk-vars.outputs.ess-region }}
+      tf-state: ${{ steps.gen-tf-name.outputs.tf-file}}
+    steps:
+      - id: set-elk-vars
+        if: github.event_name == 'workflow_dispatch'
+        env:
+          ELK_VERSION: ${{ github.event.inputs.elk-stack-version }}
+          ESS_REGION: ${{ github.event.inputs.ess-region }}
+        run: |
+          echo "stack-version=${{ env.ELK_VERSION }}" >> $GITHUB_OUTPUT
+          echo "ess-region=${{ env.ESS_REGION }}" >> $GITHUB_OUTPUT
+      - id: gen-tf-name
+        run: |
+          echo "tf-file=terraform_$(date +'%Y-%m-%d_%H-%M-%S').tfstate" >> $GITHUB_OUTPUT
+  
+  Deploy:
+    needs: ["Setup"]
+    runs-on: ubuntu-20.04
+    timeout-minutes: 15
+    defaults:
+      run:
+        working-directory: ${{ env.WORKING_DIR }}
+    env:
+      TF_VAR_stack_version: ${{ needs.Setup.outputs.stack-version }}
+      TF_VAR_ess_region: ${{ needs.Setup.outputs.ess-region }}
+      TF_STATE_FILE: ${{ needs.Setup.outputs.tf-state }}      
+    steps:
+      - name: Check out the repo
+        uses: actions/checkout@v3
+
+      - name: Init Hermit
+        run: ./bin/hermit env -r >> $GITHUB_ENV
+        working-directory: ./
+
+      - name: Configure AWS credentials
+        uses: aws-actions/configure-aws-credentials@v1-node16
+        with:
+          aws-access-key-id: ${{ env.AWS_ACCESS_KEY_ID }}
+          aws-secret-access-key: ${{ env.AWS_SECRET_ACCESS_KEY }}
+          aws-region: ${{ env.AWS_REGION }}
+
+      - name: Terraform Init
+        run: terraform init -no-color
+
+      - name: Terraform Validate
+        run: terraform validate -no-color
+
+      - name: Provision Elastic Cloud
+        id: apply
+        run: |
+          terraform apply --auto-approve \
+            -target ${{ env.EC_DEPLOYMENT }} \
+            -state=${{ env.TF_STATE_FILE}}
+
+      - name: Upload tf state
+        if: success()
+        run: |
+          aws s3 cp "${{ env.TF_STATE_FILE }}" "s3://tf-state-bucket-test-infra/${{ env.TF_STATE_FILE }}"
```

Comment:
Can you elaborate why do we need this file in S3? We should be able to list those environments by a naming convention. Also the cleanup step always runs, do you still worry about an edge case that will keep some environments up?

Reply:
There are several use cases for this workflow.
1. Weekly sanity checks -> in this case we need to provision infrastructure -> execute tests -> destroy infrastructure.
2. Provision infrastructure for some reasons: before QA cycle, once a week, etc. in this case destroy infrastructure is not relevant because you need this environment for manual testing or investigation of something and therefore you should decide when to destroy. To do that we need to save terraform state file.